File: solutions/2026-09-10/java/2026-09-10-count-nodes-equal-to-average-of-subtree.java
```java
/**
 * Problem: Count Nodes Equal to Average of Subtree
 * Problem Summary: Count nodes in a binary tree where the node's value equals the average of its subtree.
 * Link: https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/
 *
 * Approach:
 * We can solve this problem using a Depth-First Search (DFS) approach. For each node, we need to calculate the sum of all node values in its subtree and the count of nodes in its subtree.
 *
 * We define a helper function, `dfs`, that will traverse the tree. This function will return an array of two integers:
 * 1. The sum of values in the subtree rooted at the current node.
 * 2. The count of nodes in the subtree rooted at the current node.
 *
 * During the DFS traversal, for each node:
 * 1. Recursively call `dfs` on its left and right children to get their subtree sums and counts.
 * 2. Calculate the current node's subtree sum by adding its own value to the sums returned from its children.
 * 3. Calculate the current node's subtree count by adding 1 (for the current node) to the counts returned from its children.
 * 4. Compute the average of the current node's subtree: `average = current_subtree_sum / current_subtree_count`.
 * 5. If the current node's value is equal to this computed average, increment a global counter.
 * 6. Return the `[current_subtree_sum, current_subtree_count]` for the parent node to use.
 *
 * The base case for the recursion is when a node is null, in which case we return `[0, 0]` (sum and count).
 *
 * Time Complexity: O(N), where N is the number of nodes in the tree. Each node is visited exactly once during the DFS traversal.
 * Space Complexity: O(H), where H is the height of the tree. This is due to the recursion stack. In the worst case (a skewed tree), H can be N, leading to O(N) space. In the best case (a balanced tree), H is logN, leading to O(logN) space.
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // Counter to store the number of nodes that satisfy the condition.
    private int count = 0;

    /**
     * Public method to initiate the count of nodes equal to their subtree average.
     * @param root The root of the binary tree.
     * @return The number of nodes whose value is equal to the average of its subtree.
     */
    public int averageOfSubtree(TreeNode root) {
        // Start the DFS traversal from the root.
        dfs(root);
        // Return the final count.
        return count;
    }

    /**
     * Helper DFS function to calculate subtree sum and count, and check the condition.
     * @param node The current node being visited.
     * @return An array of two integers: [sum of subtree values, count of nodes in subtree].
     */
    private int[] dfs(TreeNode node) {
        // Base case: If the node is null, return [0, 0] for sum and count.
        if (node == null) {
            return new int[]{0, 0};
        }

        // Recursively call DFS on left and right children.
        int[] leftResult = dfs(node.left);
        int[] rightResult = dfs(node.right);

        // Calculate the sum of the current node's subtree.
        // It's the sum of left subtree, right subtree, and the current node's value.
        int currentSubtreeSum = leftResult[0] + rightResult[0] + node.val;

        // Calculate the count of nodes in the current node's subtree.
        // It's the count of nodes in the left subtree, right subtree, plus the current node itself.
        int currentSubtreeCount = leftResult[1] + rightResult[1] + 1;

        // Calculate the average of the current subtree, rounding down.
        int average = currentSubtreeSum / currentSubtreeCount;

        // Check if the current node's value is equal to the computed average.
        if (node.val == average) {
            // If they are equal, increment the global counter.
            count++;
        }

        // Return the sum and count for the current subtree to the parent.
        return new int[]{currentSubtreeSum, currentSubtreeCount};
    }

    // Definition for a binary tree node (provided by LeetCode, included for completeness if running locally).
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
```

File: solutions/2026-09-10/python/2026-09-10-count-nodes-equal-to-average-of-subtree.py
```python
# Problem: Count Nodes Equal to Average of Subtree
# Link: https://leetcode.com/problems/count-nodes-equal-to-average-of-subtree/
# Approach:
# We can use a Depth-First Search (DFS) approach to traverse the tree.
# For each node, we need to calculate the sum of its subtree and the count of nodes in its subtree.
# A post-order traversal is suitable because we need the information from the children before processing the parent.
# The DFS function will return a tuple: (sum_of_subtree, count_of_nodes_in_subtree).
# During the traversal, for each node, we calculate its subtree sum and count.
# Then, we compute the average (sum // count). If the node's value equals this average, we increment a counter.
#
# Time Complexity: O(N), where N is the number of nodes in the tree. Each node is visited exactly once.
# Space Complexity: O(H) in the best case (balanced tree) and O(N) in the worst case (skewed tree), where H is the height of the tree. This is due to the recursion stack used by DFS.

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def averageOfSubtree(self, root: TreeNode | None) -> int:
        # Initialize a counter for nodes whose value equals the average of their subtree.
        self.count = 0

        # Helper function to perform DFS and calculate subtree sum and node count.
        def dfs(node):
            # Base case: if the node is None, return sum 0 and count 0.
            if not node:
                return 0, 0

            # Recursively call dfs for left and right children.
            left_sum, left_count = dfs(node.left)
            right_sum, right_count = dfs(node.right)

            # Calculate the sum of the current subtree (node's value + children's subtree sums).
            current_subtree_sum = node.val + left_sum + right_sum
            # Calculate the count of nodes in the current subtree (1 for the current node + children's node counts).
            current_subtree_count = 1 + left_count + right_count

            # Calculate the average of the current subtree.
            # The problem specifies rounding down to the nearest integer.
            average = current_subtree_sum // current_subtree_count

            # Check if the current node's value is equal to the calculated average.
            if node.val == average:
                # If they are equal, increment the global counter.
                self.count += 1

            # Return the sum and count of the current subtree to be used by its parent.
            return current_subtree_sum, current_subtree_count

        # Start the DFS traversal from the root.
        dfs(root)
        # Return the final count of nodes matching the criteria.
        return self.count

```